modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/loading/Load.scala

```diff
@@ -37,6 +38,46 @@ object Load {
 
   type MonadThrow[F[_]] = MonadError[F, Throwable]
 
+  /**
+   * Loading stage.
+   * Represents the finite state machine of the Loader, it can be only in one of the below stages
+   * Internal state sets the stage right before it gets executed, i.e. if it failed being in
+   * `ManifestCheck` stage it means that manifest check has failed, but it certainly has started
+   */
+  sealed trait Stage
+  object Stage {
+    /** Figure out how the migration should look like, by inspecting affected tables. First stage */
+    final case object MigrationBuild extends Stage
+    /** Pre-transaction migrations, such as ALTER COLUMN. Usually empty. Second stage */
+    final case object MigrationPre extends Stage
+    /** Checking manifest if the folder is already loaded. Third stage */
+    final case object ManifestCheck extends Stage
+    /** In-transaction migrations, such as ADD COLUMN. Fourth stage */
+    final case object MigrationIn extends Stage
+    /** Actual loading into a table. Appears for many different tables. Fifth stage */
+    final case class Loading(table: String) extends Stage
+    /** Adding manifest item, acking SQS comment. Sixth (last) stage  */
+    final case object Commiting extends Stage
+    /** Abort the loading. Can appear after any stage */
+    final case class Cancelling(reason: String) extends Stage
+  }
+
+  /** State of the loader */
+  sealed trait State
+  object State {
+    final case object Idle extends State
+    final case class Loading(folder: S3.Folder, stage: Stage) extends State
+
+    def start(folder: S3.Folder): State =
+      State.Loading(folder, Stage.MigrationBuild)
+
+    implicit val stateShow: Show[State] =
+      Show.show {
+        case Idle => "Idle"
+        case Loading(folder, stage) => s"Ongoing processing of $folder at $stage stage"
```

Comment:
It might be helpful if `Stage` also has a `Show`, using the descriptions you have put in the code comments above.

In particular... stringifying a `Cancelling` on this line is just a little bit ugly.